src/components/requester.js

```diff
@@ -28,16 +30,38 @@ module.exports = class Requester extends Monitorable(Configurable(Component)) {
     }
 
     send(...args) {
-        if (args.length == 1 || typeof args[args.length - 1] != 'function') {
-            return new Promise((resolve, reject) => {
-                this.sock.send(...args, (err, res) => {
-                    if (err) return reject(err);
-                    resolve(res);
-                });
+        const cb = args[args.length - 1];
+        const hasCallback = 'function' == typeof cb;
+
+        const sockSend = (args) => Promise.race([
+            this.sockSend(args),
+            this.createTimeout(args),
+        ]);
+
+        if (hasCallback)
+            return sockSend(args.slice(0, -1))
+                .then((res) => cb(null, res))
+                .catch(cb);
+
+        return sockSend(args);
+    }
+
+    sockSend(args) {
+        return new Promise((resolve, reject) => {
+            this.sock.send(...args, (err, res) => {
+                if (err) return reject(err);
```

Comment:
i think these lines in this callback will execute even though the timeout got rejected.

Reply:
and also, when there's a real timeout (and the other peer disconnects, etc), axon will still keep the callback — this is also in fact a memory leak. `this.sock.callbacks` is an object that keeps track of these. this `this.sock.send` assigns an ID to the callback function you pass to it, and saves it in `this.sock.callbacks`. when there's a rejection due to timeout, you should find the relevant callback in `this.sock.callbacks` and remove it.

Reply:
Now callback function is being removed when request is timed out.